src/top/run/box_generator.py

```diff
@@ -0,0 +1,227 @@
+            raise ValueError('Transformation should be a 4x4 matrix.')
+
+        new_rotation = np.matmul(transformation[:3, :3], self.rotation)
+        new_translation = transformation[:3, 3] + (np.matmul(transformation[:3, :3], self.translation))
+        
+        return Box.from_transformation(new_rotation, new_translation, self.scale)
+
+    @classmethod
+    def scaled_axis_aligned_vertices(cls, scale):
+        """
+        Returns an axis-aligned set of verticies for a box of the given scale.
+        Args:
+            scale: A 3*1 vector, specifiying the size of the box in x-y-z dimension.
+        """
+        w = scale[0] / 2.
+        h = scale[1] / 2.
+        d = scale[2] / 2.
+
+        # Define the local coordinate system, w.r.t. the center of the box
+        aabb = np.array([[0., 0., 0.], [-w, -h, -d], [-w, -h, +d],
+                         [-w, +h, -d], [-w, +h, +d], [+w, -h, -d],
+                         [+w, -h, +d], [+w, +h, -d],[+w, +h, +d]])
+        return aabb
+
+    @classmethod
+    def fit(cls, vertices):
+        """
+        Estimates a box 9-dof parameters from the given vertices.
+        Directly computes the scale of the box, then solves for orientation and translation.
+        Args:
+            vertices: A 9*3 array of points.
+                      Points are arranged as 1 + 8 (center keypoint + 8 box vertices) matrix.
+        Returns:
+            orientation: 3*3 rotation matrix.
+            translation: 3*1 translation vector.
+            scale: 3*1 scale vector.
+        """
+        orientation = np.identity(3)
+        translation = np.zeros((3, 1))
+        scale = np.zeros(3)
+
+        # The scale would remain invariant under rotation and translation.
+        # We can safely estimate the scale from the oriented box.
+        for axis in range(3):
+            for edge_id in range(4):
+            # The edges are stored in quadruples according to each axis
+                begin, end = EDGES[axis * 4 + edge_id]
+                scale[axis] += np.linalg.norm(vertices[begin, :] - vertices[end, :])
+            scale[axis] /= 4.
+
+        x = cls.scaled_axis_aligned_vertices(scale)
+        system = np.concatenate((x, np.ones((NUM_KEYPOINTS, 1))), axis=1)
+        solution, _, _, _ = lstsq(system, vertices, rcond=None)
+        orientation = solution[:3, :3].T
+        translation = solution[3, :3]
+        
+        return orientation, translation, scale
+
+    def inside(self, point):
+        """
+        Tests whether a given point is inside the box.
+        Brings the 3D point into the local coordinate of the box.
+        In the local coordinate, the looks like an axis-aligned bounding box.
+        Next checks if the box contains the point.
+        Args:
+            point: A 3*1 numpy vector.
+        Returns:
+            True if the point is inside the box, False otherwise.
+        """
+        inv_trans = np.linalg.inv(self.transformation)
```

Comment:
SE3 transformation does not need an explicit linear inverse, it would be preferable to avoid expensive `np.linalg.inv`. Also, if `inside()` is going to be called often, we should prefer to cache `inv_trans`.